subprojects/core/src/test/groovy/org/gradle/cache/internal/CacheRetainingDataFromPreviousBuildTest.groovy

```diff
@@ -67,4 +67,41 @@ class CacheRetainingDataFromPreviousBuildTest extends AbstractCrossBuildInMemory
         1 * function.apply("c") >> new Object()
         0 * function._
     }
+
+    def "does not retain values from the previous session which are not to be kept"() {
+        def function = Mock(Function)
+        def notToBeKept = new Object()
+
+        when:
+        def cache = factory.newCacheRetainingDataFromPreviousBuild { value -> value != notToBeKept }
+        cache.get("a", function)
+        cache.get("b", function)
+        cache.get("c", function)
+
+        then:
+        1 * function.apply("a") >> new Object()
+        1 * function.apply("b") >> notToBeKept
+        1 * function.apply("c") >> notToBeKept
+        0 * function._
+
+        when:
+        listenerManager.getBroadcaster(SessionLifecycleListener).beforeComplete()
+        cache.get("a", function)
+        cache.get("b", function)
+        cache.get("c", function)
+
+        then:
+        1 * function.apply("b") >> new Object()
+        1 * function.apply("c") >> notToBeKept
+        0 * function._
```

Comment:
This would read better for me if each `get()` would be accompanied by a `then:`.

Reply:
I tried it and I find it harder to read, since the overview of what is now to be kept in cache and what isn't is hidden more. I would prefer to leave it as is.

Reply:
OK.